File: internal/config/config_test.go
package config

import (
	"os"
	"testing"
)

func TestLoad_Success(t *testing.T) {
	// Set up test environment variables
	os.Setenv("GH_USER", "testuser")
	os.Setenv("GH_TOKEN", "testtoken")
	os.Setenv("GEMINI_API_KEY", "testkey")
	os.Setenv("DISCORD_WEBHOOK_URL", "https://discord.com/webhook")
	defer cleanupEnv()

	cfg, err := Load()
	if err != nil {
		t.Fatalf("Expected no error, got: %v", err)
	}

	if cfg.GitHubUser != "testuser" {
		t.Errorf("Expected GitHubUser=testuser, got: %s", cfg.GitHubUser)
	}
	if cfg.GitHubToken != "testtoken" {
		t.Errorf("Expected GitHubToken=testtoken, got: %s", cfg.GitHubToken)
	}
	if cfg.GeminiAPIKey != "testkey" {
		t.Errorf("Expected GeminiAPIKey=testkey, got: %s", cfg.GeminiAPIKey)
	}
	if cfg.DiscordWebhookURL != "https://discord.com/webhook" {
		t.Errorf(
			"Expected DiscordWebhookURL=https://discord.com/webhook, got: %s",
			cfg.DiscordWebhookURL,
		)
	}
}

func TestLoad_MissingGitHubUser(t *testing.T) {
	os.Setenv("GH_TOKEN", "testtoken")
	os.Setenv("GEMINI_API_KEY", "testkey")
	os.Setenv("DISCORD_WEBHOOK_URL", "https://discord.com/webhook")
	defer cleanupEnv()

	_, err := Load()
	if err == nil {
		t.Fatal("Expected error for missing GH_USER, got nil")
	}

	expected := "GH_USER environment variable is required"
	if err.Error() != expected {
		t.Errorf("Expected error: %s, got: %s", expected, err.Error())
	}
}

func TestLoad_MissingGitHubToken(t *testing.T) {
	os.Setenv("GH_USER", "testuser")
	os.Setenv("GEMINI_API_KEY", "testkey")
	os.Setenv("DISCORD_WEBHOOK_URL", "https://discord.com/webhook")
	defer cleanupEnv()

	_, err := Load()
	if err == nil {
		t.Fatal("Expected error for missing GH_TOKEN, got nil")
	}

	expected := "GH_TOKEN environment variable is required"
	if err.Error() != expected {
		t.Errorf("Expected error: %s, got: %s", expected, err.Error())
	}
}

func TestLoad_MissingGeminiAPIKey(t *testing.T) {
	os.Setenv("GH_USER", "testuser")
	os.Setenv("GH_TOKEN", "testtoken")
	os.Setenv("DISCORD_WEBHOOK_URL", "https://discord.com/webhook")
	defer cleanupEnv()

	_, err := Load()
	if err == nil {
		t.Fatal("Expected error for missing GEMINI_API_KEY, got nil")
	}

	expected := "GEMINI_API_KEY environment variable is required"
	if err.Error() != expected {
		t.Errorf("Expected error: %s, got: %s", expected, err.Error())
	}
}

func TestLoad_MissingDiscordWebhookURL(t *testing.T) {
	os.Setenv("GH_USER", "testuser")
	os.Setenv("GH_TOKEN", "testtoken")
	os.Setenv("GEMINI_API_KEY", "testkey")
	defer cleanupEnv()

	_, err := Load()
	if err == nil {
		t.Fatal("Expected error for missing DISCORD_WEBHOOK_URL, got nil")
	}

	expected := "DISCORD_WEBHOOK_URL environment variable is required"
	if err.Error() != expected {
		t.Errorf("Expected error: %s, got: %s", expected, err.Error())
	}
}

func cleanupEnv() {
	os.Unsetenv("GH_USER")
	os.Unsetenv("GH_TOKEN")
	os.Unsetenv("GEMINI_API_KEY")
	os.Unsetenv("DISCORD_WEBHOOK_URL")
}

File: src/config/config.test.ts
import { afterEach, beforeEach, describe, expect, it } from "bun:test";
import { ConfigError, loadConfig } from "./index";

describe("loadConfig", () => {
  const originalEnv = { ...Bun.env };

  beforeEach(() => {
    Bun.env.GH_USER = "testuser";
    Bun.env.GH_TOKEN = "testtoken";
    Bun.env.GEMINI_API_KEY = "testapikey";
    Bun.env.DISCORD_WEBHOOK_URL = "https://discord.com/webhook";
  });

  afterEach(() => {
    Object.assign(Bun.env, originalEnv);
  });

  it("should load config with all required variables", () => {
    const config = loadConfig();

    expect(config.ghUser).toBe("testuser");
    expect(config.ghToken).toBe("testtoken");
    expect(config.geminiApiKey).toBe("testapikey");
    expect(config.discordWebhookUrl).toBe("https://discord.com/webhook");
  });

  it("should use default gemini model when not specified", () => {
    Bun.env.GEMINI_MODEL = undefined;
    const config = loadConfig();

    expect(config.geminiModel).toBe("gemini-2.5-flash");
  });

  it("should use custom gemini model when specified", () => {
    Bun.env.GEMINI_MODEL = "gemini-pro";
    const config = loadConfig();

    expect(config.geminiModel).toBe("gemini-pro");
  });

  it("should throw ConfigError when GH_USER is missing", () => {
    Bun.env.GH_USER = undefined;

    expect(() => loadConfig()).toThrow(ConfigError);
    expect(() => loadConfig()).toThrow("GH_USER");
  });

  it("should throw ConfigError when GH_TOKEN is missing", () => {
    Bun.env.GH_TOKEN = undefined;

    expect(() => loadConfig()).toThrow(ConfigError);
    expect(() => loadConfig()).toThrow("GH_TOKEN");
  });

  it("should throw ConfigError when GEMINI_API_KEY is missing", () => {
    Bun.env.GEMINI_API_KEY = undefined;

    expect(() => loadConfig()).toThrow(ConfigError);
    expect(() => loadConfig()).toThrow("GEMINI_API_KEY");
  });

  it("should throw ConfigError when DISCORD_WEBHOOK_URL is missing", () => {
    Bun.env.DISCORD_WEBHOOK_URL = undefined;

    expect(() => loadConfig()).toThrow(ConfigError);
    expect(() => loadConfig()).toThrow("DISCORD_WEBHOOK_URL");
  });

  it("should list all missing variables in error message", () => {
    Bun.env.GH_USER = undefined;
    Bun.env.GH_TOKEN = undefined;

    expect(() => loadConfig()).toThrow("GH_USER");
    expect(() => loadConfig()).toThrow("GH_TOKEN");
  });
});
